state/azure/cosmosdb/cosmosdb.go

```diff
@@ -377,95 +346,103 @@ func (c *StateStore) Delete(req *state.DeleteRequest) error {
 				return err
 			}
-			upsertOperation := CosmosOperation{
-				Item: item,
-				Type: upsertOperationType,
+
+			if req.ETag != nil && *req.ETag != "" {
+				etag := azcore.ETag(*req.ETag)
+				options.IfMatchETag = &etag
+			}
+			if req.Options.Concurrency == state.FirstWrite && (req.ETag == nil || *req.ETag == "") {
+				u, err := uuid.NewRandom()
+				if err != nil {
+					return err
+				}
+				options.IfMatchETag = ptr.Of(azcore.ETag(u.String()))
 			}
-			operations = append(operations, upsertOperation)
+
+			marsh, err := json.Marshal(doc)
+			if err != nil {
+				return err
+			}
+			batch.UpsertItem(marsh, nil)
+			numOperations++
 		} else if o.Operation == state.Delete {
 			req := o.Request.(state.DeleteRequest)
 
-			deleteOperation := CosmosOperation{
-				Item: CosmosItem{
-					ID:           req.Key,
-					Value:        "", // Value does not need to be specified
-					PartitionKey: partitionKey,
-				},
-				Type: deleteOperationType,
+			if req.ETag != nil && *req.ETag != "" {
+				etag := azcore.ETag(*req.ETag)
+				options.IfMatchETag = &etag
+			}
+			if req.Options.Concurrency == state.FirstWrite && (req.ETag == nil || *req.ETag == "") {
+				u, err := uuid.NewRandom()
+				if err != nil {
+					return err
+				}
+				options.IfMatchETag = ptr.Of(azcore.ETag(u.String()))
 			}
-			operations = append(operations, deleteOperation)
+
+			batch.DeleteItem(req.Key, options)
+			numOperations++
 		}
 	}
 
-	c.logger.Debugf("#operations=%d,partitionkey=%s", len(operations), partitionKey)
+	c.logger.Debugf("#operations=%d,partitionkey=%s", numOperations, partitionKey)
 
-	var retString string
-
-	// The stored procedure throws if it failed, which sets err to non-nil.  It doesn't return anything else.
-	err := retryOperation(func() error {
-		innerErr := c.client.ExecuteStoredProcedure(
-			c.getSprocLink(storedProcedureName),
-			[...]interface{}{operations},
-			&retString,
-			documentdb.PartitionKey(partitionKey),
-		)
-		if innerErr != nil {
-			if isTooManyRequestsError(innerErr) {
-				return innerErr
-			}
-			return backoff.Permanent(innerErr)
-		}
-		return nil
-	}, func(err error, d time.Duration) {
-		c.logger.Warnf("CosmosDB state store Multi request failed: %v; retrying in %s", err, d)
-	}, 20*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	batchResponse, err := c.client.ExecuteTransactionalBatch(ctx, batch, nil)
+	cancel()
 	if err != nil {
-		c.logger.Debugf("error=%e", err)
 		return err
 	}
-
+	if batchResponse.Success {
+		// Transaction succeeded
+		// We can inspect the individual operation results
+		for index, operation := range batchResponse.OperationResults {
+			c.logger.Debugf("Operation %v completed with status code %d", index, operation.StatusCode)
+		}
+	} else {
+		// Transaction failed, look for the offending operation
+		for index, operation := range batchResponse.OperationResults {
+			if operation.StatusCode != http.StatusTooManyRequests {
```

Comment:
Why this here?

Reply:
I understand that this was here from before, but before this was in a retry block... Do we need this check now? What if the error is a 429, what should we do?